File: melbourne/graphics/utilities.py
from enum import Enum, auto
from pathlib import Path

import webcolors
from cachetools import TTLCache, cached
from skia import (
    Canvas,
    Color,
    FilterQuality,
    Font,
    Image,
    Paint,
    Rect,
    TextBlob,
)


def get_text_width_height(font: Font, text: str) -> tuple[float, float]:
    """Get the width and height for a text, using the font for metric calculations"""
    width = font.measureText(text)
    height = font.getSpacing()
    return width, height


def hex_to_rgb(hex: str) -> Color:
    """Convert a color from HEX representation to RGB (in Skia.Color)"""
    rgb = webcolors.hex_to_rgb(hex)
    return Color(rgb.red, rgb.green, rgb.blue)


def draw_rectangle(
    canvas: Canvas,
    point: tuple[int, int],
    width: int,
    height: int,
    # Technically, skia.Color is just an int
    fill_color: int | None = None,
    stroke_color: int | None = None,
    stroke_width: float | None = None,
) -> None:
    """Draw a rectangle, which is optionally filled in, or has a border"""
    x, y = point
    rect = Rect(x, y, x + width, y + height)

    if fill_color:
        paint = Paint(AntiAlias=True, Color=fill_color, Style=Paint.Style.kFill_Style)
        canvas.drawRect(rect, paint)

    if stroke_color and stroke_width:
        paint = Paint(
            AntiAlias=True,
            Color=stroke_color,
            Style=Paint.Style.kStroke_Style,
            StrokeWidth=stroke_width,
        )
        canvas.drawRect(rect, paint)


class TextAlignment(Enum):
    LEFT = auto()
    CENTER = auto()


def draw_text(
    canvas: Canvas,
    point: tuple[float, float],
    text: str,
    font: Font,
    color: Color,
    alignment: TextAlignment = TextAlignment.LEFT,
) -> None:
    """Draw text using the defined font and color"""
    x, y = point
    blob = TextBlob(text, font)
    paint = Paint(AntiAlias=True, Color=color)

    if alignment == TextAlignment.CENTER:
        # Horizontally center the text around x
        width = font.measureText(text)
        x -= width / 2.0

    # Vertically center the text around y
    metrics = font.getMetrics()
    y -= (metrics.fAscent + metrics.fDescent) / 2.0

    canvas.drawTextBlob(blob, x, y, paint)


def draw_line(
    canvas: Canvas,
    starting_point: tuple[float, float],
    ending_point: tuple[float, float],
    color: Color,
    stroke_width: float,
) -> None:
    """Draw a line with the specified color and stroke width"""
    x0, y0 = starting_point
    x1, y1 = ending_point
    paint = Paint(
        AntiAlias=True,
        Color=color,
        StrokeWidth=stroke_width,
        Style=Paint.Style.kStroke_Style,
    )
    canvas.drawLine(x0, y0, x1, y1, paint)


@cached(cache=TTLCache(ttl=60, maxsize=1024))
def load_image(file_path: Path, width: float) -> Image:
    """Load an image. The image is cached for 60 seconds so that subsequent load attempts for the current scoreboard are cached."""
    with open(file_path, "rb") as f:
        data = f.read()
    image = Image.MakeFromEncoded(data)

    orig_width = image.width()
    orig_height = image.height()
    aspect_ratio = orig_height / orig_width
    height = int(width * aspect_ratio)

    resized_image = image.resize(
        int(width),
        int(height),
        FilterQuality.kHigh_FilterQuality,
    )
    return resized_image


def draw_image(
    canvas: Canvas,
    image: Image,
    x: float,
    y: float,
    stroke_color: int | None = None,
    stroke_width: float | None = None,
):
    """Draw an image, with an optional border"""
    if not (stroke_color and stroke_width):
        canvas.drawImage(image, x, y)
    else:
        # Draw the image centered within the border
        src_rect = Rect.MakeWH(image.width(), image.height())
        dst_rect = Rect.MakeXYWH(x, y, image.width(), image.height())
        canvas.drawImageRect(image, src_rect, dst_rect)

        # Draw the border for the image
        paint = Paint(
            AntiAlias=True,
            Color=stroke_color,
            Style=Paint.Style.kStroke_Style,
            StrokeWidth=stroke_width,
        )
        border_rect = Rect.MakeXYWH(
            x,
            y,
            image.width(),
            image.height(),
        )
        canvas.drawRect(border_rect, paint)

File: Melbourne/Scoreboard/Utilities.cs
namespace Melbourne;

using System;
using System.IO;
using Microsoft.Extensions.Caching.Memory;
using SkiaSharp;

public static class Utilities
{
    private static readonly MemoryCache ImageCache =
        new(new MemoryCacheOptions { SizeLimit = 1024 });
    
    /* -------------------------------------------------------------
     * Text helpers
     * ------------------------------------------------------------- */

    public static (float Width, float Height) GetTextWidthHeight(
        SKFont font,
        string text)
    {
        var width = font.MeasureText(text);
        var height = font.Spacing;
        return (width, height);
    }

    /* -------------------------------------------------------------
     * Color helpers
     * ------------------------------------------------------------- */

    public static SKColor HexToRgb(string hex)
    {
        if (hex.StartsWith("#"))
        {
            hex = hex[1..];
        }

        var r = Convert.ToByte(hex[..2], 16);
        var g = Convert.ToByte(hex.Substring(2, 2), 16);
        var b = Convert.ToByte(hex.Substring(4, 2), 16);

        return new SKColor(r, g, b);
    }

    /* -------------------------------------------------------------
     * Drawing helpers
     * ------------------------------------------------------------- */

    public static void DrawRectangle(
        SKCanvas canvas,
        (int X, int Y) point,
        int width,
        int height,
        SKColor? fillColor = null,
        SKColor? strokeColor = null,
        float? strokeWidth = null)
    {
        var rect = new SKRect(
            point.X,
            point.Y,
            point.X + width,
            point.Y + height);

        if (fillColor.HasValue)
        {
            using var paint = new SKPaint
            {
                IsAntialias = true,
                Color = fillColor.Value,
                Style = SKPaintStyle.Fill
            };

            canvas.DrawRect(rect, paint);
        }

        if (strokeColor.HasValue && strokeWidth.HasValue)
        {
            using var paint = new SKPaint
            {
                IsAntialias = true,
                Color = strokeColor.Value,
                Style = SKPaintStyle.Stroke,
                StrokeWidth = strokeWidth.Value
            };

            canvas.DrawRect(rect, paint);
        }
    }

    public enum TextAlignment
    {
        Left,
        Center
    }

    public static void DrawText(
        SKCanvas canvas,
        (float X, float Y) point,
        string text,
        SKFont font,
        SKColor color,
        TextAlignment alignment = TextAlignment.Left)
    {
        using var paint = new SKPaint
        {
            IsAntialias = true,
            Color = color
        };

        var x = point.X;
        var y = point.Y;

        if (alignment == TextAlignment.Center)
        {
            var width = font.MeasureText(text);
            x -= width / 2f;
        }

        font.GetFontMetrics(out var metrics);
        y -= (metrics.Ascent + metrics.Descent) / 2f;

        canvas.DrawText(text, x, y, font, paint);
    }

    public static void DrawLine(
        SKCanvas canvas,
        (float X, float Y) start,
        (float X, float Y) end,
        SKColor color,
        float strokeWidth)
    {
        using var paint = new SKPaint
        {
            IsAntialias = true,
            Color = color,
            StrokeWidth = strokeWidth,
            Style = SKPaintStyle.Stroke
        };

        canvas.DrawLine(start.X, start.Y, end.X, end.Y, paint);
    }
    
    public static void DrawImage(
        SKCanvas canvas,
        SKBitmap image,
        float x,
        float y,
        SKColor? strokeColor = null,
        float? strokeWidth = null)
    {
        if (!(strokeColor.HasValue && strokeWidth.HasValue))
        {
            canvas.DrawBitmap(image, x, y);
            return;
        }

        var rect = new SKRect(
            x,
            y,
            x + image.Width,
            y + image.Height);

        canvas.DrawBitmap(image, rect);

        using var paint = new SKPaint
        {
            IsAntialias = true,
            Color = strokeColor.Value,
            Style = SKPaintStyle.Stroke,
            StrokeWidth = strokeWidth.Value
        };

        canvas.DrawRect(rect, paint);
    }
    
    /* -------------------------------------------------------------
     * Image helpers
     * ------------------------------------------------------------- */

    public static SKBitmap? LoadImage(
        string filePath,
        float width)
    {
        var cacheKey = $"{filePath}:{width}";
        if (ImageCache.TryGetValue(cacheKey, out SKBitmap? cached))
        {
            return cached;
        }

        using var stream = File.OpenRead(filePath);
        using var original = SKBitmap.Decode(stream);

        var aspectRatio = (float)original.Height / original.Width;
        var height = (int)(width * aspectRatio);

        var resized = original.Resize(
            new SKImageInfo((int)width, height),
            SKFilterQuality.High);

        // Cache images for 60 seconds to speed up scoreboard generation for the current run
        ImageCache.Set(
            cacheKey,
            resized,
            new MemoryCacheEntryOptions
            {
                AbsoluteExpirationRelativeToNow = TimeSpan.FromSeconds(60),
                Size = 1
            });

        return resized;
    }
}